### PR TITLE
Add options for syncing buffers between GPUs

### PR DESCRIPTION
**Pull request by edpizzi**
Summary:
This adds two buffer synchronization options to ClassificationTask.

Synchronizing buffers between GPUs allows stronger guarantees that evals in the training run will be reproducible from the model snapshot, since both parameters and buffers are the same. Without this, each GPU's BatchNorm stats (as a motivating example) may be slightly different, producing different results from the snapshot (which reflects GPU 0's BN statistics).

The first option is `broadcast_buffers`, which simply enables `DistributedDataParallel`'s `broadcast_buffers` option, broacasting GPU 0's buffers to all GPUs every forward pass. This solves the problem, but incurs synchronization overhead during training.

The second option, `broadcast_buffers_before_eval` synchronizes buffers only once per epoch, between train and eval phases. This solves buffer synchronization for BatchNorm statistics, which are only used for evaluation, without incurring training overhead.

Note: This PyTorch issue tracks adding a DistributeDataParallel method for explicit buffer synchronization. https://github.com/pytorch/pytorch/issues/30718

Differential Revision: D18795308

### CODE REVIEW

File: classy_vision/tasks/classification_task.py [hunk 1]
```diff
@@ -707,9 +738,28 @@ def _set_model_train_mode(self):
...
+        This provides the eval time stability of `broadcast_buffers`,
+        especially for BatchNorm statistics which are not used during training,
+        without the latency of `broadcast_buffers`.
+        """
+        if self.distributed_model is None:
+            return
+        buffers = list(self.base_model.buffers())
+        if len(buffers) > 0:
+            logging.info("Synchronizing buffers before evaluation.")
+            _broadcast_coalesced(
```
Comment (edpizzi) [inline classy_vision/tasks/classification_task.py]: Note: this only handles the case where each GPU has a separate controlling process, which I believe is appropriate given Classy's process model, but I just wanted to flag it.

My understanding is that if a process controls multiple GPUs, we'd need a second intra-process synchronization across GPUs controlled by this process (as DistirbutedDataParallel's `_sync_params` method has).

We can avoid using an undocumented / private API here at the cost of some performance by broadcasting buffers individually. Since this is a once-per-epoch call, performance is not that important, so I'm open to either approach.
